content/blog/elm-at-noredink/index.md

```diff
@@ -0,0 +1,445 @@
+in the world. It serves millions of teachers and students every day and our
+frontend is exclusively written in Elm. In this post, we will explore the
+structure of our application and the patterns that we use to stay sane.
+
+One of the most common questions I get about Elm is "But does it scale? And
+if so, how". Well, I hope that this post can become a comprehensive answer
+to that question.
+
+## Some stats first
+
+We have a glorious monorepo which contains all the services that power our
+infrastructure. Our main application (aka the `monolith`) is written in
+Rails and has a `ui` folder with all of our Elm code.
+
+```bash
+$ find monolith/ui/src -name "*.elm" | xargs cat | wc -l
+275082
+
+$ find monolith/ui/tests -name "*.elm" | xargs cat | wc -l
+209107
+```
+
+As you can see, we write loads of tests. As much as the Elm compiler gives
+you that wonderful bliss of refactoring, we still want to make sure that
+the code that we deploy is as correct as it can be.
+
+In the somewhat famous article ["What to know before debating type
+systems"](https://cdsmith.wordpress.com/2011/01/09/an-old-article-i-wrote/),
+two approaches are given which can help us writing correct programs:
+
+- **Testing**: establishes upper bounds on correctness
+- **Proof**: establishes lower bounds on correctness
+
+In our case, static types are _lightweight_ proofs:
+
+> An important point here is that static typing does not preclude proving
+> correctness in the traditional way, nor testing the program. It is a
+> technique to handle those cases in which testing might be guaranteed to
+> succeed so they don’t need testing; and similarly, to provide a basis
+> from which the effort of manual proof can be saved for those truly
+> challenging areas in which it is necessary.
+
+## Rails conventions
+
+Following the Rails architecture, each resource is managed by their own
+controller. So when the teacher wants to manage their classes:
+
+- they visit the `/teach/classes` URL
+- that route is managed by the `Teach::ClassesController` controller
+- that controller will fetch stuff from the database, clean it up and load a view
+
+That view looks like this:
+
+```haml
+- content_for :title, "Manage Classes | NoRedInk"
+
+- content_for(:javascript) do
+  = javascript_include_tag 'webpack/teach/classes'
+
+= elm_mount(@page_data, 'teach-classes-elm')
+```
+
+What is that `elm_mount` helper? We defined it to be like this:
+
+```ruby
+def elm_mount(flags, prefix = 'elm')
+  tag.div(id: prefix + '-flags',
+          class: "elm-flags",
+          data: {flags: flags.to_json}) +
+    tag.div(id: prefix + '-host')
+end
+```
+
+In short, we need some conventions for the name of the two divs:
+- the one that contains the flags that we pass from Rails to Elm
+- the one that will contain the actual Elm app when it's mounted
+
+The javascript file that we include is the result of the compilation of the
+Elm app. We use Coffeescript for making an entrypoint which looks like
```

Comment:
💭  I think it might be better to skip the mention to coffee and use the premades endpoint as an example of a pure JS set up:
  - even though coffee is used on the majority of our pages, we are moving away from it
  - more readers will be familiar with JS than coffee, and adding a new language might make the picture more complex (which is unnecessary given that there's nothing special about coffee at this point IMO)

Reply:
(`Page/Admin/Curriculum/Premades/index.js` shows a similar initialization snippet written in JS)

Reply:
Right! I wanted to do that, but it looked a bit ugly due to the imported paths `../../../../` I guess I can just clean it up :)